HW2/P3/mandelbrot.pyx

```diff
@@ -26,21 +23,79 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
        #     v3 = AVX.mul(v1, v2)
        #
        # You may find the numpy.real() and numpy.imag() fuctions helpful.
+       int i, j, k, iter
+       #np.complex64_t c, z
+       AVX.float8 real_pic_c, imag_pic_c, real_pic_z, imag_pic_z, cz_pic,
+       AVX.float8 mag, tmp_num_a, tmp_num_b, mask, mask2, real_pic_z_tmp, cnt_iteration, cnt_finally, cnt_iteration_filt 
+       float out_vals[8] 
+
+       
 
     assert in_coords.shape[1] % 8 == 0, "Input array must have 8N columns"
     assert in_coords.shape[0] == out_counts.shape[0], "Input and output arrays must be the same size"
     assert in_coords.shape[1] == out_counts.shape[1],  "Input and output arrays must be the same size"
 
     with nogil:
-        for i in range(in_coords.shape[0]):
-            for j in range(in_coords.shape[1]):
-                c = in_coords[i, j]
-                z = 0
+        for i in prange(in_coords.shape[0], schedule = 'static', chunksize = 1, num_threads = 4):
+        #for i in range(in_coords.shape[0]):
+            for j in range(0,in_coords.shape[1],8):
+                #c = in_coords[i, j]
+                #z = 0
+                #from back to start 
+                real_pic_c = AVX.make_float8( in_coords[i, j + 7].real, in_coords[i, j + 6].real,
+                                              in_coords[i, j + 5].real, in_coords[i, j + 4].real,
+                                              in_coords[i, j + 3].real, in_coords[i, j + 2].real,
+                                              in_coords[i, j + 1].real, in_coords[i, j + 0].real )
+
+                imag_pic_c = AVX.make_float8( in_coords[i, j + 7].imag, in_coords[i, j + 6].imag,
+                                              in_coords[i, j + 5].imag, in_coords[i, j + 4].imag,
+                                              in_coords[i, j + 3].imag, in_coords[i, j + 2].imag,
+                                              in_coords[i, j + 1].imag, in_coords[i, j + 0].imag )
+
+                real_pic_z = AVX.float_to_float8(0.0)
+                imag_pic_z = AVX.float_to_float8(0.0)
+                cz_pic = AVX.float_to_float8(2.0)
+                cnt_finally = AVX.float_to_float8(0.0)
+                
                 for iter in range(max_iterations):
-                    if magnitude_squared(z) > 4:
-                        break
-                    z = z * z + c
-                out_counts[i, j] = iter
+                #    if mag_squared(z) > 4:
+                #        break
+                #    z = z * z + c
+                #out_counts[i, j] = iter
+                    # set tmp_num_a
+                    tmp_num_a = AVX.mul(real_pic_z, real_pic_z) 
+                    # set tmp_num_b          
+                    tmp_num_b = AVX.mul(imag_pic_z, imag_pic_z) 
+                    # set magnitude          
+                    mag  = AVX.sqrt( AVX.add( tmp_num_a, tmp_num_b ) ) 
```

Comment:
You could instead check if mag^2<4. It would save an operation, but I doubt it makes much difference either way.